cdap-app-fabric/src/main/java/io/cdap/cdap/internal/capability/CapabilityStatusStore.java

```diff
@@ -0,0 +1,209 @@
+        capabilityOperationMap.put(capability, capabilityOperationRecord);
+      });
+      //query all capabilities
+      StructuredTable capabilityTable = context.getTable(StoreDefinition.CapabilitiesStore.CAPABILITIES);
+      CloseableIterator<StructuredRow> capabilityResultIterator = capabilityTable.scan(Range.all(), Integer.MAX_VALUE);
+      capabilityResultIterator.forEachRemaining(structuredRow -> {
+        String capability = structuredRow.getString(StoreDefinition.CapabilitiesStore.NAME_FIELD);
+        CapabilityStatusRecord capabilityStatusRecord = new CapabilityStatusRecord(
+          capability,
+          CapabilityStatus
+            .valueOf(structuredRow.getString(StoreDefinition.CapabilitiesStore.STATUS_FIELD).toUpperCase()),
+          GSON
+            .fromJson(structuredRow.getString(StoreDefinition.CapabilitiesStore.CONFIG_FIELD), CapabilityConfig.class));
+        //add to result and remove from operations map
+        capabilityRecordMap
+          .put(capability,
+               new CapabilityRecord(capabilityStatusRecord, capabilityOperationMap.remove(capability)));
+      });
+      //add the remaining operations to result
+      capabilityOperationMap.keySet().forEach(capability -> capabilityRecordMap
+        .put(capability, new CapabilityRecord(null, capabilityOperationMap.get(capability))));
+      return capabilityRecordMap;
+    }, IOException.class);
+  }
+
+  /**
+   * Add or update capability
+   *
+   * @param capability
+   * @param status
+   * @throws IOException
+   */
+  @Override
+  public void addOrUpdateCapability(String capability, CapabilityStatus status,
+                                    CapabilityConfig config) throws IOException {
+    TransactionRunners.run(transactionRunner, context -> {
+      StructuredTable capabilityTable = context.getTable(StoreDefinition.CapabilitiesStore.CAPABILITIES);
+      Collection<Field<?>> fields = new ArrayList<>();
+      fields.add(Fields.stringField(StoreDefinition.CapabilitiesStore.NAME_FIELD, capability));
+      fields.add(Fields.stringField(StoreDefinition.CapabilitiesStore.STATUS_FIELD, status.name().toLowerCase()));
+      fields.add(Fields.stringField(StoreDefinition.CapabilitiesStore.CONFIG_FIELD, GSON.toJson(config)));
+      fields.add(Fields.longField(StoreDefinition.CapabilitiesStore.UPDATED_TIME_FIELD, System.currentTimeMillis()));
+      capabilityTable.upsert(fields);
+    }, IOException.class);
+  }
+
+  /**
+   * Delete capability
+   *
+   * @param capability
+   * @throws IOException
+   */
+  @Override
+  public void deleteCapability(String capability) throws IOException {
+    TransactionRunners.run(transactionRunner, context -> {
+      StructuredTable capabilityTable = context.getTable(StoreDefinition.CapabilitiesStore.CAPABILITIES);
+      Collection<Field<?>> fields = new ArrayList<>();
+      fields.add(Fields.stringField(StoreDefinition.CapabilitiesStore.NAME_FIELD, capability));
```

Comment:
You can just use `Collections.singleton(Fields.stirngField(....))` instead of creating an array list.

Reply:
Done